### ISSUE TITLE
Source control module fails with `Error: spawn ENOMEM`

### ISSUE THREAD

**Issue by svillar** [microsoft/vscode#63866]
Version: 1.29.1
Commit: bc24f98b5f70467bc689abf41cc5550ca637088e
Date: 2018-11-15T19:07:43.495Z
Electron: 2.0.12
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: x64

Steps to Reproduce:
1. Perform some actions with the source control module (git repository)
2. Try to change branch using the button on the button on the bottom left corner (some other actions cause the error too but I consistently get errors doing that)

The git command fails with the following error:
```
Error: spawn ENOMEM
	at _errnoException (util.js:1024:11)
	at ChildProcess.spawn (internal/child_process.js:323:11)
	at exports.spawn (child_process.js:514:9)
	at Object.exports.fork (child_process.js:115:10)
	at Promise (/usr/share/code/resources/app/extensions/json-language-features/client/dist/jsonMain.js:1:115339)
	at new Promise (<anonymous>)
	at _getServerWorkingDir.then.i (/usr/share/code/resources/app/extensions/json-language-features/client/dist/jsonMain.js:1:114945)
	at <anonymous>
```

Does this issue occur when all extensions are disabled?: Yes

**Comment by joaomoreno** [comment]
`ENOMEM` means you're out of memory. Do you have limited RAM?

**Comment by joaomoreno** [comment]
Which platform are you on?

**Comment by svillar** [comment]
> Which platform are you on?

Linux, Debian unstable in particular

**Comment by joaomoreno** [comment]
Does everything else in Code work?

**Comment by svillar** [comment]
> Does everything else in Code work?

Yep, actually works pretty good, congrats for the great job.

The git log does not show anything special, is there any environment variable or something like that that I could use to enable verbose debugging in order to know what's going on there?

**Comment by joaomoreno** [comment]
Did you see this? https://stackoverflow.com/questions/26193654/node-js-catch-enomem-error-thrown-after-spawn

**Comment by svillar** [comment]
> Did you see this? https://stackoverflow.com/questions/26193654/node-js-catch-enomem-error-thrown-after-spawn

Not sure what you mean. I'm probably hitting that same issue indeed but as I mentioned my system is not in a OOM situation and I do have swap contrary to what the other report says.

But yeah the error seems to be related to a malfunctioning spawn implementation.

**Comment by joaomoreno** [comment]
Can you try spawning git from a simple Node JS script?

```
const cp = require('child_process');
console.log(cp.spawnSync('git', ['--version']));
```

**Comment by svillar** [comment]
> Can you try spawning git from a simple Node JS script?
> 
> ```js
> console.log(require('child_process').spawnSync('git', ['--version'], { encoding: 'utf8' }).stdout);
> ```

Works fine here, it prints a bunch of envvars among other stuff and finishes with an _undefined_. I'll execute it whenever I get the error if you want

**Comment by joaomoreno** [comment]
Oh so you don't get the error all the time?

**Comment by svillar** [comment]
> Oh so you don't get the error all the time?

I used to get it all the time.

Now I think I have something. I increased the limit for watched files for [some other reason](https://code.visualstudio.com/docs/setup/linux#_visual-studio-code-is-unable-to-watch-for-file-changes-in-this-large-workspace-error-enospc) and since then it has been working completely fine.

**Comment by joaomoreno** [comment]
@bpasero Can this error occur also when the open file limit is hit?

**Comment by bpasero** [comment]
@joaomoreno this is the first time I hear of `ENOMEM`, typically the watcher one is `ENOSPC`

**Comment by svillar** [comment]
> @bpasero Can this error occur also when the open file limit is hit?

I don't think so. I also tried to increase it to the limit (and I'm far from that value) and the error is still there.

**Comment by joaomoreno** [comment]
This is pretty odd. Do you know of any other weird thing happening in your system as processes are spawned?

**Comment by svillar** [comment]
No issue at all. I use it everyday under heavy loads (browser engines development) without any problem so far.

**Comment by joaomoreno** [comment]
What git version do you have?

**Comment by svillar** [comment]
```$ git --version
git version 2.19.1```

**Comment by joaomoreno** [comment]
Does git 2.20.0 fix it?

**Comment by svillar** [comment]
> Does git 2.20.0 fix it?

Nop. I've updated and I've just got the same ENOMEM error. There not seem to be any OOM situation:
```
$ free
              total        used        free      shared  buff/cache   available
Mem:       16272864    11388200     1086116      700352     3798548     3842216
Swap:       3903484     3711088      192396
```

**Comment by svillar** [comment]
BTW I'm just trying to switch branches, this is what the git log says `git checkout -q 1370279`.

**Comment by tlaziuk** [comment]
I've had the same error, when I was trying to add any file via the `Source Control`, but it stopped occuring when I've disabled the GitLens (v9.4.1) extension.

The error:
```
[2019-02-05 17:04:55:679] git — -c, core.quotepath=false, -c, color.ui=false, blame, --root, --incremental, --, <confidential>,   cwd='<confidential>'

  
Error: spawn ENOMEM
```

``` sh
# code --version
1.30.2
61122f88f0bf01e2ac16bdb9e1bc4571755f5bd8
x64
```

``` sh
# node -e "console.log(require('child_process').spawnSync('git', ['--version'], { encoding: 'utf8' }).stdout);"
git version 2.20.1
```

**Comment by svillar** [comment]
> I've had the same error, when I was trying to add any file via the `Source Control`, but it stopped occuring when I've disabled the GitLens (v9.4.1) extension.

I tried that too but just worked for a while. I think that the `ENOMEM`is a legit error. I found out that vscode was consuming most of my 16Gb most of the time so I guess it's quite sensible to think that there is no bug there but just an excesive memory usage. I've upgraded to 32Gb recently (due to some other reason) and I haven't experienced it so far.
